src/node/txreconciliation.cpp

```diff
@@ -142,9 +250,104 @@ class TxReconciliationTracker::Impl
+    {
+        auto fanout_candidates = m_tx_fanout_targets_cache_data.find(wtxid);
+        if (fanout_candidates != m_tx_fanout_targets_cache_data.end()) {
+            return std::binary_search(fanout_candidates->second.begin(), fanout_candidates->second.end(), peer_id);
+        }
+
+        // We use the pre-determined randomness to give a consistent result per transaction,
+        // thus making sure that no transaction gets "unlucky" if every per-peer roll fails.
+        CSipHasher deterministic_randomizer{m_deterministic_randomizer};
+        deterministic_randomizer.Write(wtxid.ToUint256());
+
+        // The algorithm works as follows. We iterate through the peers (of a given direction)
+        // hashing them with the given wtxid, and sort them by this hash.
+        // We then consider top `limit` peers to be low-fanout flood targets.
+        // The randomness should be seeded with wtxid to return consistent results for every call.
+
+        // The integral part of `limit` is accounted in the higher 32 bits of the second element.
+        // The fractional part of `limit` is stored in the lower 32 bits, and then we check
+        // whether adding a random lower 32-bit value (first element) would end up modifying
+        // the higher bits.
+        const size_t targets_size = ((deterministic_randomizer.Finalize() & 0xFFFFFFFF) + uint64_t(limit * 0x100000000)) >> 32;
+
+        std::vector<std::pair<uint64_t, NodeId>> best_peers;
+        best_peers.reserve(m_states.size());
+
+        for (const auto& indexed_state : m_states) {
+            const auto cur_state = std::get_if<TxReconciliationState>(&indexed_state.second);
+            if (cur_state && cur_state->m_we_initiate == we_initiate) {
+                uint64_t hash_key = CSipHasher(deterministic_randomizer).Write(cur_state->m_k0).Finalize();
+                best_peers.emplace_back(hash_key, indexed_state.first);
+            }
+        }
+
+        // Sort by the assigned key.
+        std::sort(best_peers.begin(), best_peers.end());
+        best_peers.resize(targets_size);
+
+        std::vector<NodeId> new_fanout_candidates;
+        new_fanout_candidates.reserve(targets_size);
+        for_each(best_peers.begin(), best_peers.end(),
+                 [&new_fanout_candidates](auto& keyed_peer) { new_fanout_candidates.push_back(keyed_peer.second); });
+
+        // Sort by NodeId.
+        std::sort(new_fanout_candidates.begin(), new_fanout_candidates.end());
+        bool found = std::binary_search(new_fanout_candidates.begin(), new_fanout_candidates.end(), peer_id);
+
+        // If the cache is full, make room for the new entry.
+        if (m_tx_fanout_targets_cache_order.size() == FANOUT_TARGETS_PER_TX_CACHE_SIZE) {
+            auto expired_tx = m_tx_fanout_targets_cache_order.front();
+            m_tx_fanout_targets_cache_data.erase(expired_tx);
+            m_tx_fanout_targets_cache_order.pop_front();
```

Comment:
In the eventuality of an influx of inbound transactions, faster than we can be flush out them to low-fanout flooding peers, my understanding of dropping the upfront wtxid candidate, we would keep propagating this transaction only according to tx-relay policy and connection state of other peers (not this `NodeId` anymore).

I understand we’re fanning out only to outbound peers (`m_tx_fanout_targets_cache_data` doc), though here it’s more a dependency on the perfomance capabilities of the full-node itself (i.e how fast you process `vInv(MSG_WTX`) and how fast you-reannounce them to downstream peers if valid). To interferes with a transaction propagation, assuming a non-listening node, an attacker would have to be puppet or compromise all our low-fanout outbound peers, I think ? Obviously more outbound peers would make things better on this front, which should be allowed by Erlay tx-relay bandwidth savings.